File: python/src/azoth/eos/reference/freezing_point.py
"""``eos.freezing_point`` - the temperature at which a pure substance's solid appears.

The Python reference, the mirror of ``crates/azoth-eos/src/freezing_point.rs``:

```text
solve  (g_fluid(T, P) - g_solid(T, P)) / (R T) = 0   for T
```

Two things the Rust module's docstring states and a reader of *this* file needs in the same
breath, because they are where a port goes wrong silently:

* **the solid is calibrated, not raw.** ``eos.parahydrogen_solid_phase`` is the thesis
  equation as it stands and its reference is fixed elsewhere - by NeqSim's
  ``SystemLeachmanEos``, which shifts it to the para-Leachman liquid at the triple point.
  Unshifted it misses the triple point's own definition, the two Gibbs energies equal there,
  by ``3.12`` J/mol.
* **the fluid's root is NeqSim's rule**: a gas below the triple-point pressure and a liquid at
  and above it, which are different states at the same temperature and pressure.
"""

from __future__ import annotations

import math
from collections.abc import Callable
from typing import Any

from azoth.core.errors import (
    AzothError,
    InvalidInputError,
    OutOfRangeError,
    SolverNotConvergedError,
)
from azoth.core.range import apply_checks, checks_for
from azoth.core.result import FreezingPointResult, Phase
from azoth.core.units import Q, from_si, input_to_si
from azoth.core.warnings import Warning
from azoth.eos.reference import hydrogen_phase as fluid
from azoth.eos.reference.parahydrogen_solid_phase import parahydrogen_solid_phase
from azoth.eos.reference.pt_flash import pt_flash
from azoth.eos.reference.tp_solid_flash import METHANE_NEVER_FREEZES, _tabulated_solid_fugacity

MODEL_ID = "eos.freezing_point"

#: NeqSim's ``ThermodynamicConstantsInterface.R``, which its residual divides by. **Not** the
#: equation's own: the two reference equations carry ``8.31451`` and ``8.3144621``, each a
#: rounded value of its own era, and this is the third.
R = 8.3144621

#: The tolerance on the dimensionless residual and on the bracket.
RESIDUAL_TOLERANCE = 1.0e-10
BRACKET_TOLERANCE = 1.0e-12

#: How far the search may walk, in kelvin, and how many steps it may take.
MINIMUM_TEMPERATURE = 0.5
MAXIMUM_TEMPERATURE = 5000.0
SPAN_GROWTH = 1.8
MAXIMUM_BRACKET_STEPS = 50
MAXIMUM_SOLVER_STEPS = 100

#: The triple point, and the enthalpy of fusion there: the coordinates the calibration is
#: measured at. Para-hydrogen's, from NeqSim's ``ParaHydrogenSolidHelmholtzEquation``.
TRIPLE_POINT_TEMPERATURE = 13.8033
TRIPLE_POINT_PRESSURE = 7042.0
TRIPLE_POINT_ENTHALPY_OF_FUSION = 118.0

#: The substances this has a solid equation for. One name, because
#: ``thermo/util/solid/`` holds two equations and the other is argon's.
SUBSTANCES = ("para-hydrogen",)


def _solid_state(t: float, p: float) -> tuple[float, float]:
    """The raw solid's molar Gibbs energy and entropy at a state, in J/mol and J/(mol*K)."""
    state = parahydrogen_solid_phase(T=from_si(t, "K"), P=from_si(p, "Pa"))
    return state.g.to("J/mol").magnitude, state.s.to("J/(mol*K)").magnitude


def _fluid_state(t: float, p: float, root: str) -> tuple[float, float]:
    """The fluid's molar Gibbs energy and entropy, on the root named, at a state."""
    density = fluid._solve_density(t, p, "para") if root == "gas" else _dense_density(t, p)
    properties = fluid._properties(t, density, "para")
    return properties["g"], properties["s"]


def _dense_density(t: float, p: float) -> float:
    density = fluid._solve_density_dense(t, p, "para")
    if density is None:
        raise OutOfRangeError(
            "P",
            p,
            "the freezing search asks for the fluid's dense root, and this pressure has none "
            "at this temperature within the range the equation is fitted to",
        )
    return density


def calibration() -> tuple[float, float]:
    """The two constants NeqSim's ``SystemLeachmanEos`` computes at the triple point.

    ``(gibbs_shift, entropy_shift)``, in J/mol and J/(mol*K). Measurements of the two models
    rather than tunables: the first comes out ``-3.122683``, which is the offset the capture
    measures between NeqSim's solid and the raw one.
    """
    g_liquid, s_liquid = _fluid_state(TRIPLE_POINT_TEMPERATURE, TRIPLE_POINT_PRESSURE, "liquid")
    g_raw, s_raw = _solid_state(TRIPLE_POINT_TEMPERATURE, TRIPLE_POINT_PRESSURE)
    return (
        g_liquid - g_raw,
        s_liquid - s_raw - TRIPLE_POINT_ENTHALPY_OF_FUSION / TRIPLE_POINT_TEMPERATURE,
    )


def calibrated_solid_gibbs(t: float, p: float, shifts: tuple[float, float]) -> float:
    """The solid's molar Gibbs energy at a state, on the reference the liquid is on."""
    gibbs_shift, entropy_shift = shifts
    g_raw, _ = _solid_state(t, p)
    # A shift of `-S (T - T_tp) + G` in the Helmholtz energy is the same shift in the Gibbs
    # energy: it carries no volume, so `g = a + P v` passes it through.
    return g_raw - entropy_shift * (t - TRIPLE_POINT_TEMPERATURE) + gibbs_shift


def residual(t: float, p: float, root: str, shifts: tuple[float, float]) -> float:
    """The dimensionless residual the solve drives to zero."""
    g_fluid, _ = _fluid_state(t, p, root)
    return (g_fluid - calibrated_solid_gibbs(t, p, shifts)) / (R * t)


def _solve(
    residual_fn: Callable[[float], float], start: float, algorithm: Any
) -> tuple[float, int, float]:
    """The temperature a residual crosses zero at, by NeqSim's own walk.

    The search starts where the algorithm says and expands a span either side of it: the answer
    is a property of the substance and the pressure, so the start is a path and not a state. A
    trial the equation cannot evaluate is skipped rather than fatal, which is NeqSim's
    ``evaluateResidualIfValid`` - near the triple point one side of a span can ask for a dense
    root that does not exist.

    Raises:
        SolverNotConvergedError: if no span brackets a sign change, or the bracket collapses
            without reaching the tolerance.
    """
    start_residual = residual_fn(start)
    if abs(start_residual) < RESIDUAL_TOLERANCE:
        return start, 1, start_residual

    span = max(start * 0.01, 0.1)
    low, low_residual = start, start_residual
    # The bracket's upper side is only read through its sign; it starts at the same point
    # because a bracket needs two sides and either can move first.
    high, high_residual = start, start_residual
    iterations = 1
    bracketed = False
    for _ in range(MAXIMUM_BRACKET_STEPS):
        candidate_low = max(start - span, MINIMUM_TEMPERATURE)
        candidate_high = min(start + span, MAXIMUM_TEMPERATURE)
        # **Any error, not just an out-of-range one**, because that is what the Rust kernel
        # catches: its `if let Ok(value) = residual(...)` discards every `Err`. The tabulated
        # route's residual is a whole flash, so it can fail as `SolverNotConverged` too.
        try:
            low, low_residual = candidate_low, residual_fn(candidate_low)
            iterations += 1
        except AzothError:
            pass
        try:
            high, high_residual = candidate_high, residual_fn(candidate_high)
            iterations += 1
        except AzothError:
            pass
        if low_residual * high_residual <= 0.0:
            bracketed = True
            break
        span *= SPAN_GROWTH

    if not bracketed:
        raise _not_converged(iterations, start_residual, algorithm)

    for _ in range(MAXIMUM_SOLVER_STEPS):
        mid = 0.5 * (low + high)
        value = residual_fn(mid)
        iterations += 1
        if abs(value) < RESIDUAL_TOLERANCE or high - low < BRACKET_TOLERANCE:
            return mid, iterations, value
        if low_residual * value <= 0.0:
            high = mid
        else:
            low, low_residual = mid, value

    raise _not_converged(iterations, 0.5 * (low + high), algorithm)


def _log(x: float) -> float:
    """``log``, giving NaN outside its domain as Java and Rust do."""
    try:
        return math.log(x)
    except ValueError:
        return math.nan


def _tabulated_residual(
    mixture: Any, z: list[float], index: int, solid: str, eos: str, t: float, p: float
) -> float:
    """NeqSim's ``calculateLogEquilibriumResidual`` outside its Helmholtz branch.

    ``residual = ln z_k - ln( sum_p beta_p phi_solid_k / phi_kp )``, the multiphase appearance
    condition over the phases the flash found. The largest logarithm is taken out before
    exponentiating: the terms are ratios of fugacity coefficients spanning twenty orders of
    magnitude and a direct sum underflows.
    """
    if not z[index] > 0.0:
        raise InvalidInputError(
            "solid",
            f"`{solid}` has no overall mole fraction, and a solid cannot appear from a "
            f"substance the feed does not have",
        )
    flash = pt_flash(mixture, from_si(t, "K"), from_si(p, "Pa"), z)
    # **Methane never freezes in NeqSim**: `ComponentSolid.fugcoef` returns `1e30` for it before
    # any arithmetic, so its residual has no sign change and the search refuses.
    if solid.strip().lower() == "methane":
        phi_solid = METHANE_NEVER_FREEZES
    else:
        phi_solid = _tabulated_solid_fugacity(
            mixture, index, solid, t, p, from_si(t, "K"), from_si(p, "Pa"), eos
        )
    # **Not `math.log`: Python raises where Java and Rust give NaN.** NeqSim's
    # `ComponentSolid.fugcoef` can return a negative coefficient - the class makes one for
    # water by dividing two signed expressions - and Rust's `f64::ln` makes that a `NaN` the
    # search then skips or refuses on, while `math.log` aborts the twin at the same state. The
    # same helper, and the same reason, as `saft_vr_mie_phase._log`.
    ln_phi_solid = _log(phi_solid)

    # The fluid's phases, as `(amount, ln phi of this component)`. **The match is on the
    # phase and not on whether a beta came back**, which is the Rust kernel's order and not an
    # equivalent: a flash can report `all_liquid` with a beta still set, and the two orders
    # then read a different phase's coefficient.
    terms: list[float] = []
    if flash.phase is Phase.TWO_PHASE:
        raw = flash.beta if flash.beta is not None else 0.0
        terms.append(_log(1.0 - raw) + ln_phi_solid - flash.ln_phi_liquid[index])
        terms.append(_log(raw) + ln_phi_solid - flash.ln_phi_vapour[index])
    elif flash.phase is Phase.ALL_LIQUID:
        terms.append(ln_phi_solid - flash.ln_phi_liquid[index])
    else:
        terms.append(ln_phi_solid - flash.ln_phi_vapour[index])
    maximum = max(terms)
    if not math.isfinite(maximum):
        raise OutOfRangeError(
            "P",
            p,
            "no phase of the flashed fluid has a finite fugacity contribution at this state",
        )
    scaled = sum(math.exp(term - maximum) for term in terms)
    return _log(z[index]) - maximum - _log(scaled)


def freezing_point(components: list[str], z: list[float], solid: str, P: Q) -> FreezingPointResult:
    """The temperature at which a fluid's solid-forming substance freezes.

    Two routes, as NeqSim has two: ``para-hydrogen`` against its calibrated solid Helmholtz
    equation, every other substance against the tabulated solid whose residual is the multiphase
    appearance condition.

    Raises:
        InvalidInputError: if ``z`` does not match ``components``, if ``solid`` is not one of
            them, if ``P`` is not positive, or if the tabulated route's candidate carries no
            melt data.
        OutOfRangeError: from a trial's state.
        SolverNotConvergedError: if no candidate's residual brackets a sign change.
    """
    from azoth import _models_gen

    spec = _models_gen.model(MODEL_ID)
    checks = checks_for(spec)
    warnings: list[Warning] = []
    p_si = input_to_si(spec, "P", P)
    apply_checks(checks.on_input, {"P": p_si}.get, warnings)

    if len(components) != len(z):
        raise InvalidInputError(
            "z",
            f"{len(components)} components and {len(z)} mole fractions; a candidate's own "
            f"fraction is what decides whether it can appear",
        )
    names = [name.strip().lower() for name in components]
    wanted = solid.strip().lower()
    if wanted not in names:
        raise InvalidInputError(
            "solid",
            f"{solid!r} is not one of the fluid's components, so it is not a substance whose "
            f"freezing point this can solve for",
        )
    index = names.index(wanted)

    algorithm = spec["algorithm"]
    start = float(algorithm.get("initial_temperature", 14.0))

    if wanted == "para-hydrogen":
        # NeqSim's own rule, from `initializeCoexistingFluidPhase`: the fluid is a gas below
        # the triple-point pressure and a liquid at and above it.
        root = "gas" if p_si < TRIPLE_POINT_PRESSURE else "liquid"
        shifts = calibration()
        temperature, iterations, value = _solve(
            lambda t: residual(t, p_si, root, shifts), start, algorithm
        )
    else:
        from azoth.eos import components as databank

        mixture, _ = databank.mixture_of(components, eos="srk")
        temperature, iterations, value = _solve(
            lambda t: _tabulated_residual(mixture, z, index, solid, "srk", t, p_si),
            start,
            algorithm,
        )

    return FreezingPointResult(
        temperature=from_si(temperature, "K"),
        component=components[index],
        iterations=iterations,
        residual=value,
        warnings=tuple(warnings),
    )


def _not_converged(iterations: int, value: float, algorithm: dict[str, object]) -> Exception:
    return SolverNotConvergedError(
        iterations=iterations,
        residual=value,
        tolerance=float(str(algorithm["tolerance"])),
    )


__all__ = ["freezing_point"]

File: crates/azoth-eos/src/freezing_point.rs
//! `eos.freezing_point` - the temperature at which a fluid's solid-forming substance appears.
//!
//! NeqSim's `FreezingPointTemperatureFlash`, both of the routes it has, chosen by an
//! `instanceof` upstream and by the candidate's name here:
//!
//! - **[`SolidRoute::Helmholtz`]** - the substance has a solid reference equation
//!   (`thermo/util/solid/`), and the residual is the molar Gibbs difference:
//!
//!   ```text
//!   solve  (g_fluid(T, P) - g_solid(T, P)) / (R T) = 0   for T
//!   ```
//!
//! - **[`SolidRoute::Tabulated`]** - every other substance, against the tabulated solid built
//!   from COMP.csv's melting point, heat of fusion, heat capacities and densities. Its residual
//!   is [`tabulated_residual`]'s multiphase appearance condition, which is a log-sum-exp over
//!   the fluid's own phases rather than a difference of two Gibbs energies.
//!
//! Both are wrapped by the same [`solve`], which is NeqSim's own walk: a span either side of a
//! starting temperature, expanded until it brackets a sign change, then bisected.
//!
//! # The calibration, which is not in the solid
//!
//! NeqSim's `ParaHydrogenSolidHelmholtzEquation` is the **raw** thesis equation, and the
//! reference it is reported on is fixed by its *system*:
//! `SystemLeachmanEos.createCalibratedParaHydrogenSolidEquation` shifts it to the
//! para-Leachman liquid at the triple point,
//!
//! ```text
//! gibbsShift   = g_liquid(T_tp, P_tp) - g_raw_solid(T_tp, P_tp)
//! entropyShift = s_liquid(T_tp, P_tp) - s_raw_solid(T_tp, P_tp) - H_fus / T_tp
//! ```
//!
//! so that solid and liquid meet there - which is what a triple point *is*, and what the raw
//! equation alone does not do. This module applies it, because it is the thing that compares
//! the two phases; `eos.parahydrogen_solid_phase` stays raw and says so.
//!
//! # Which root of the fluid
//!
//! The flash re-initialises its fluid phase as a gas below the triple-point pressure and a
//! liquid above, and takes the root that follows. That is what [`FluidRoot`] reproduces: a
//! pure substance's isotherm crosses a pressure three times, and the two roots are different
//! states at the same temperature and pressure.

use azoth_core::units::{Pressure, kelvins, pascals};
use azoth_core::{AzothError, Result, apply_checks};

use crate::leachman::{self, HydrogenType};
use crate::mixture::Mixture;
use crate::model_gen;
use crate::parahydrogen_solid;
use crate::results::FreezingPointResult;

/// NeqSim's `ThermodynamicConstantsInterface.R`, which its residual divides by. **Not** the
/// equation's own `R`: the two reference equations here carry their own rounded constants, and
/// this is the third.
pub const R: f64 = 8.314_462_1;

/// The tolerance on the dimensionless residual, and on the bracket.
const RESIDUAL_TOLERANCE: f64 = 1.0e-10;
const BRACKET_TOLERANCE: f64 = 1.0e-12;

/// How far the search may walk, in kelvin, and how many steps it may take.
const MINIMUM_TEMPERATURE: f64 = 0.5;
const MAXIMUM_TEMPERATURE: f64 = 5000.0;
const SPAN_GROWTH: f64 = 1.8;
const MAXIMUM_BRACKET_STEPS: usize = 50;
const MAXIMUM_SOLVER_STEPS: usize = 100;

/// The solid's reference shift at the triple point, in J/mol and J/(mol.K).
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct Calibration {
    /// Added to the raw solid's Gibbs energy, and carried with `T - T_tp` by the entropy
    /// shift.
    pub gibbs_shift: f64,
    /// What the raw solid's entropy is short of the liquid's at the triple point.
    pub entropy_shift: f64,
}

/// Which root of the fluid's isotherm the equilibrium is with.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum FluidRoot {
    /// Below the triple-point pressure.
    Gas,
    /// At and above it.
    Liquid,
}

/// The calibration NeqSim's `SystemLeachmanEos` applies, computed the way it computes it.
///
/// The liquid is `para`-hydrogen's dense root at the triple point - the fluid the shift is
/// measured against - and the solid is this crate's raw equation at the same state. The two
/// constants are measurements of those two models rather than tunables: at the triple point
/// they come out `-3.122683` J/mol and a positive entropy shift whose size is fixed by the
/// fusion enthalpy.
#[must_use]
pub fn calibration() -> Calibration {
    let t_tp = parahydrogen_solid::TRIPLE_POINT_TEMPERATURE;
    let p_tp = parahydrogen_solid::TRIPLE_POINT_PRESSURE;

    let rho = leachman::solve_density_dense(t_tp, p_tp, HydrogenType::Para)
        .expect("the para-hydrogen liquid root exists at the triple point");
    let liquid = leachman::properties(t_tp, rho, HydrogenType::Para);
    let raw_solid = parahydrogen_solid::properties(t_tp, p_tp);

    Calibration {
        gibbs_shift: liquid.g - raw_solid.g,
        entropy_shift: liquid.s
            - raw_solid.s
            - parahydrogen_solid::TRIPLE_POINT_ENTHALPY_OF_FUSION / t_tp,
    }
}

/// The solid's molar Gibbs energy at a state, on the reference the liquid is on.
#[must_use]
pub fn calibrated_solid_gibbs(t: f64, p: f64, calibration: &Calibration) -> f64 {
    let raw = parahydrogen_solid::properties(t, p);
    // `g = a + Pv` gains the shift directly, and the entropy shift's own temperature
    // dependence comes with it: a shift of `-S (T - T_tp) + G` in the Helmholtz energy is
    // `-S (T - T_tp) + G` in the Gibbs energy too, since the shift carries no volume.
    raw.g - calibration.entropy_shift * (t - parahydrogen_solid::TRIPLE_POINT_TEMPERATURE)
        + calibration.gibbs_shift
}

/// The dimensionless residual the solve drives to zero, and the one NeqSim's
/// `calculateLogEquilibriumResidual` returns on this route.
///
/// # Errors
/// * [`AzothError::OutOfRange`] if the fluid's dense root does not exist at `t`.
pub fn residual(t: f64, p: f64, root: FluidRoot, calibration: &Calibration) -> Result<f64> {
    let rho = match root {
        FluidRoot::Gas => leachman::solve_density(t, p, HydrogenType::Para),
        FluidRoot::Liquid => {
            leachman::solve_density_dense(t, p, HydrogenType::Para).ok_or_else(|| {
                AzothError::out_of_range(
                    "P",
                    p,
                    "the freezing search asks for the fluid's dense root, and this pressure \
                     has none at this temperature within the range the equation is fitted to",
                )
            })?
        }
    };
    let fluid = leachman::properties(t, rho, HydrogenType::Para);
    let solid = calibrated_solid_gibbs(t, p, calibration);
    Ok((fluid.g - solid) / (R * t))
}

/// Which of NeqSim's two solid routes a substance is solved on.
///
/// NeqSim dispatches on the *system*: `FreezingPointTemperatureFlash` asks whether its solid
/// phase is a `PhaseSolidHelmholtzEos`, which `SystemLeachmanEos` and `SystemSolidHelmholtzEos`
/// build and every other system does not. The set of substances with a Helmholtz solid here is
/// the one `thermo/util/solid/` gives an equation for, so this is derived from the name rather
/// than declared - and it agrees with NeqSim's because NeqSim's own systems are what carry those
/// equations.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum SolidRoute {
    /// A solid reference equation: the calibrated para-hydrogen one, whose residual is the
    /// molar Gibbs difference.
    Helmholtz,
    /// The tabulated solid: `ComponentSolid.fugcoef2` over the databank's melting point, heat
    /// of fusion, heat capacities and density correlations, and the residual is the multiphase
    /// appearance condition over the fluid's own phases.
    Tabulated,
}

/// The route a named substance is solved on.
#[must_use]
pub fn route_for(solid: &str) -> SolidRoute {
    if matches!(normalised(solid).as_str(), "para-hydrogen") {
        SolidRoute::Helmholtz
    } else {
        SolidRoute::Tabulated
    }
}

/// A name as the databank spells it: trimmed and lower-cased.
fn normalised(name: &str) -> String {
    name.trim().to_lowercase()
}

/// The dimensionless residual the tabulated route drives to zero, for one candidate.
///
/// NeqSim's `calculateLogEquilibriumResidual`, outside its Helmholtz branch:
///
/// ```text
/// residual = ln z_k - ln( sum_p beta_p phi_solid_k / phi_kp )
/// ```
///
/// **It is the multiphase appearance condition, and the log-sum-exp is why it is written that
/// way.** A fluid can be several phases at once and the solid meets all of them, so the sum is
/// over the phases the flash found, weighted by their amounts. NeqSim takes the maximum
/// logarithm out before exponentiating, which is not a rearrangement for its own sake: the
/// terms are ratios of fugacity coefficients across twenty orders of magnitude, and a direct
/// sum underflows to zero and makes the logarithm infinite.
///
/// # Errors
/// * [`AzothError`] from the flash, from the solid coefficient, or from a candidate the fluid
///   does not contain.
fn tabulated_residual(
    mixture: &Mixture,
    z: &[f64],
    index: usize,
    solid: &str,
    eos: &str,
    t: f64,
    p: f64,
) -> Result<f64> {
    // A `NaN` fraction is as much a refusal as a zero one, and `<=` alone would let it past.
    if z[index] <= 0.0 || !z[index].is_finite() {
        return Err(AzothError::invalid_input(
            "solid",
            format!(
                "`{solid}` has no overall mole fraction, and a solid cannot appear from a \
                 substance the feed does not have"
            ),
        ));
    }
    let flash = crate::pt_flash::pt_flash(mixture, kelvins(t), pascals(p), z)?;
    // **Methane never freezes in NeqSim.** `ComponentSolid.fugcoef` opens with
    // `if (componentName.equals("methane")) return 1e30;`, and that is the entry point the
    // solid phase calls - so a methane candidate's solid is infinitely volatile, its residual
    // has no sign change, and the search refuses rather than reporting a freezing point. This
    // is that guard.
    let phi_solid = if normalised(solid) == "methane" {
        crate::tp_solid_flash::METHANE_NEVER_FREEZES
    } else {
        crate::tp_solid_flash::tabulated_solid_fugacity(
            mixture,
            index,
            solid,
            kelvins(t),
            pascals(p),
            eos,
        )?
    };
    let ln_phi_solid = phi_solid.ln();

    // The fluid's phases, as `(amount, ln phi of this component)`. A two-phase flash has two;
    // a single-phase one has one, carrying the whole feed.
    let mut phases: Vec<(f64, f64)> = Vec::new();
    match flash.phase {
        crate::results::Phase::TwoPhase => {
            let beta = flash.beta.unwrap_or(0.0);
            phases.push((1.0 - beta, flash.ln_phi_liquid[index]));
            phases.push((beta, flash.ln_phi_vapour[index]));
        }
        crate::results::Phase::AllLiquid => phases.push((1.0, flash.ln_phi_liquid[index])),
        _ => phases.push((1.0, flash.ln_phi_vapour[index])),
    }

    let terms: Vec<f64> = phases
        .iter()
        .filter(|(amount, _)| *amount > 0.0)
        .map(|(amount, ln_phi)| amount.ln() + ln_phi_solid - ln_phi)
        .collect();
    let maximum = terms.iter().copied().fold(f64::NEG_INFINITY, f64::max);
    if !maximum.is_finite() {
        return Err(AzothError::out_of_range(
            "P",
            p,
            "no phase of the flashed fluid has a finite fugacity contribution at this state",
        ));
    }
    let scaled: f64 = terms.iter().map(|term| (term - maximum).exp()).sum();
    Ok(z[index].ln() - maximum - scaled.ln())
}

/// Where a solve starts, and how far it may walk.
#[derive(Debug, Clone, Copy)]
struct Search {
    start: f64,
}

/// The temperature a residual crosses zero at, by NeqSim's own walk.
///
/// The search starts where the algorithm says and expands a span either side of it: the answer
/// is a property of the substance and the pressure, so the start is a path and not a state. A
/// trial the equation cannot evaluate is skipped rather than fatal, which is NeqSim's
/// `evaluateResidualIfValid` - near a triple point one side of a span can ask for a root that
/// does not exist.
///
/// # Errors
/// * [`AzothError::SolverNotConverged`] if no span brackets a sign change, or the bracket
///   collapses without reaching the tolerance.
fn solve<F>(residual: F, search: Search, tolerance: f64) -> Result<(f64, f64, u32)>
where
    F: Fn(f64) -> Result<f64>,
{
    let start = search.start;
    let start_residual = residual(start)?;
    if start_residual.abs() < RESIDUAL_TOLERANCE {
        return Ok((start, start_residual, 1));
    }

    let mut span = (start * 0.01).max(0.1);
    let (mut lo, mut lo_residual) = (start, start_residual);
    // The bracket's upper side is only read through its residual's sign, which the walk
    // refreshes below; it starts at the same point as the lower one because a bracket needs
    // two sides and either can move first.
    let (mut hi, mut hi_residual) = (start, start_residual);
    let mut iterations = 1;
    let mut bracketed = false;
    for _ in 0..MAXIMUM_BRACKET_STEPS {
        let candidate_lo = (start - span).max(MINIMUM_TEMPERATURE);
        let candidate_hi = (start + span).min(MAXIMUM_TEMPERATURE);
        if let Ok(value) = residual(candidate_lo) {
            lo = candidate_lo;
            lo_residual = value;
            iterations += 1;
        }
        if let Ok(value) = residual(candidate_hi) {
            hi = candidate_hi;
            hi_residual = value;
            iterations += 1;
        }
        if lo_residual * hi_residual <= 0.0 {
            bracketed = true;
            break;
        }
        span *= SPAN_GROWTH;
    }
    if !bracketed {
        return Err(AzothError::SolverNotConverged {
            iterations,
            residual: start_residual,
            tolerance,
        });
    }

    for _ in 0..MAXIMUM_SOLVER_STEPS {
        let mid = 0.5 * (lo + hi);
        let value = residual(mid)?;
        iterations += 1;
        if value.abs() < RESIDUAL_TOLERANCE || hi - lo < BRACKET_TOLERANCE {
            return Ok((mid, value, iterations));
        }
        if lo_residual * value <= 0.0 {
            // The upper side moves; its residual is not stored because the next step
            // recomputes one at the midpoint and compares against `lo_residual` alone.
            hi = mid;
        } else {
            lo = mid;
            lo_residual = value;
        }
    }

    Err(AzothError::SolverNotConverged {
        iterations,
        residual: 0.5 * (lo + hi),
        tolerance,
    })
}

/// The freezing-point temperature of a fluid's solid-forming substance.
///
/// Two routes, as NeqSim has two: para-hydrogen is solved against its calibrated solid
/// reference equation, and every other substance against the tabulated solid built from its
/// melting point, heat of fusion and densities.
///
/// **A fluid may have more than one candidate and the highest freezing point is the answer.**
/// NeqSim loops the components the caller enabled a solid check for and keeps the maximum,
/// because a fluid at a temperature where any one of its substances freezes has frozen; the
/// result names the one that set it. `solid` is that set, and naming one substance is the
/// single-candidate case.
///
/// # Errors
/// * [`AzothError::InvalidInput`] if `z` does not match `components`, if `solid` is not one of
///   them, if `P` is not positive, or if the tabulated route's candidate carries no melt data.
/// * [`AzothError::OutOfRange`] from a trial's state.
/// * [`AzothError::SolverNotConverged`] if no candidate's residual brackets a sign change.
pub fn freezing_point(
    components: &[String],
    z: &[f64],
    solid: &str,
    p: Pressure,
) -> Result<FreezingPointResult> {
    let spec = &model_gen::FREEZING_POINT_SPEC;
    let mut warnings = Vec::new();
    apply_checks(
        spec.input_checks(),
        |quantity| match quantity {
            "P" => Some(p.value),
            _ => None,
        },
        &mut warnings,
    )?;

    if components.len() != z.len() {
        return Err(AzothError::invalid_input(
            "z",
            format!(
                "{} components and {} mole fractions; a candidate's own fraction is what \
                 decides whether it can appear",
                components.len(),
                z.len()
            ),
        ));
    }
    let names: Vec<String> = components.iter().map(|name| normalised(name)).collect();
    let index = names
        .iter()
        .position(|name| *name == normalised(solid))
        .ok_or_else(|| {
            AzothError::invalid_input(
                "solid",
                format!(
                    "`{solid}` is not one of the fluid's components, so it is not a substance \
                     whose freezing point this can solve for"
                ),
            )
        })?;

    let algorithm = crate::algorithm_of(spec)?;
    let search = Search {
        start: algorithm.initial_temperature.unwrap_or(14.0),
    };
    let p_pa = p.value;
    let eos = "srk";

    let (temperature, residual, iterations) = match route_for(solid) {
        SolidRoute::Helmholtz => {
            let root = if p_pa < parahydrogen_solid::TRIPLE_POINT_PRESSURE {
                FluidRoot::Gas
            } else {
                FluidRoot::Liquid
            };
            let calibration = calibration();
            solve(
                |t| residual(t, p_pa, root, &calibration),
                search,
                algorithm.tolerance,
            )?
        }
        SolidRoute::Tabulated => {
            let borrowed: Vec<&str> = components.iter().map(String::as_str).collect();
            let (mixture, _) = crate::databank::mixture_of(&borrowed, crate::Cubic::Srk, None)?;
            solve(
                |t| tabulated_residual(&mixture, z, index, solid, eos, t, p_pa),
                search,
                algorithm.tolerance,
            )?
        }
    };

    Ok(FreezingPointResult {
        temperature: kelvins(temperature),
        component: components[index].clone(),
        iterations,
        residual,
        warnings,
    })
}
